README.md
# Ad.MySports header-bidding

This project is a [prebid.js](http://prebid.org/) implementation for [ADTECH/AOL AdServer](https://oneadserver.aol.com/), including extra features as:
- auto-refresh
- [prebid analytics with GA](http://prebid.org/overview/ga-analytics.html)
- adblock tracker
- auto-detect Ad Units on DOMContentLoaded

We've took care to use javascript able to run on following browser versions:
- Chrome 49+
- Safari 9+
- IE 11+ 
- Firefox 52+
- Samsung 6.2+

We didn't test on all of these browsers versions but should work.

UPDATE 2018-05-24: we use prebid.js v1.11.0

# ONE by AOL: Dynamic Ad Call Library
Ads with asynchronous tags allow page content to be loaded independently from ad content, reducing the overall load time of the page.

[Ad Server tag library DAC.js](https://learn.oneadserver.aol.com/reader/NwjtoyhHugfpQzAulJ_o9A/5SMTKNtxD2uPB9JdxkA2Xw) (Dynamic Ad Call) is the JavaScript core component that contains the base code needed for all ad calls and additional code to, e.g., request screen properties or geo location.

# Prebid launched on load and DOMContentLoaded event
The prebid is launched on load with a predefined list of AdUnits Ids, and when DOMContentLoaded is fired the DOM is parsed to check AdUnits Ids missing on load and re-call a bid for them.
The DOMContenLoaded solution is also used when Ad Unit is loaded through an AdServer (DFP for example) and not set directly on html page.

To setup and customize the `ams-ad` each site should load a configuration with the 2 mandatory variables: hbAMS and adServer.

## hbAMS variable

hbAMS variable defined settings and AdUnits.

```
var hbAMS = {
  settings: {
    siteId: '670202',
    prebidAdUnitIds: ["6494071"],
    autoRefresh: {
      interval: 15000, // milliseconds
      minVisibility: 0.49, // range 0-1
      onlyIfBidWinner: false
    },
    gdpr: true,
    sizes: {
      desktop: [
        [970, 250],
        [970, 90]
      ],
      tablet: [
        [728, 90],
        [300, 600]
      ],
      phone: [
        [300, 250]
      ]
    },
    adUnits: [
      {
        code: "6494071",
        bounds: [
            { id: 6494072, min: 0, max: 768 },
            { id: 6494071, min: 769, max: 9999 },
        ],
        sizeid: '16',
        labelAny: ["desktop"],
        bids: []
      },
    ]
  }
};
```

## adServer variable

adServer variable defines the config of AdServer, including key/value you'd like to pass.

```
var adServer = {};

adServer.config = {
  protocol: 'https',
  server: 'secserv.adtech.de',
  network: '1502.1',
  siteid: hbAMS.settings.siteId,
  gdpr: hbAMS.settings.gdpr,
  params: {
    loc: '100',
    kvcat: '[CATEGORY NAME HERE]',
    kvurl: '[URL]',
  }
};
```

# Mandatory updates
On [v1.*](http://prebid.org/dev-docs/publisher-api-reference.html#module_pbjs.addAdUnits) Prebid.js demands **size** on adUnit i.e.:
```json
 {
    code: "6489219",
    fif: { usefif: true },
    sizes: [[728, 90], [970, 90], [970, 250]],
    bounds: [
        { id: 6493810, min: 0, max: 768 },
        { id: 6489219, min: 769, max: 9999 },
    ],
    sizeid: '225',
    bids: []
},
```

SizeMapping must be [removed](http://prebid.org/dev-docs//prebid-1.0-API.html#size-mapping-changes) from adUnits.


# Customization
## auto-refresh
The auto-refresh allows to refresh specific AdUnits each period with certain conditions. The period is defined by `interval`.

```
autoRefresh: {
    interval: 15000, // milliseconds
    minVisibility: 0.75, // range 0-1
    onlyIfBidWinner: false,
},
```
  - `minVisibility`: the min visibility required for the AdUnit on viewport to be refreshed. This should prevent to refresh Ad Units out of viewport and decrease their cpm due to low/none visibility.
  - `onlyIfBidWinner`: refresh Ad Unit only if exists a bid winner, else cancel.

Each Ad Unit to be refreshed requires the property `autoRefresh` set to true, by default this propoerty is false.

```
{
    code: 6490489,
    fif: { usefif: true },
    autoRefresh: true,
    sizes: [[300, 600], [300, 250]],
    ...
```

To prevent blink effects we recommend to only use auto-refresh on prefixed height Ad Units.

## adblock tracker
```
var defaultSettings = {
  analytics: {
      trackAdblock: true,
      trackPrebid: true,
  },
```

For more details we recommend the read of [Detecting Adblock on your site and logging it to Google Analytics](http://blog.dynamicdrive.com/detecting-adblock-on-your-site-and-logging-it-to-google-analytics/)

## prebid analytics with GA

[prebid analytics with GA](http://prebid.org/overview/ga-analytics.html)
You must to set the trackerName 
```
ga('create', 'UA-28563613-5', 'auto', 'ams');
...
pbams.que.push(function () {
      pbams.enableAnalytics({
          provider: 'ga',
          options: {
              trackerName: 'ams',
              global: 'ga',
              enableDistribution: false,
          }
      });
  });
```

# Prebid.js
## Change list of modules
To make sure the build only includes the [modules]((https://github.com/prebid/Prebid.js/tree/master/modules)) you want, you can specify the modules to be included with the --modules CLI argument

```
$ gulp build --modules=smartadserverBidAdapter,rubiconBidAdapter
```

Alternatively, a `.json` file can be specified that contains a list of modules you would like to include.

```
$ gulp build --modules=modules.json
```

With modules.json containing the following

```
[
    "consentManagement",
    "smartadserverBidAdapter",
    "rubiconBidAdapter",
    "aolBidAdapter",
    "widespaceBidAdapter",
    "improvedigitalBidAdapter",
    "criteoBidAdapter",
    "googleAnalyticsAdapter"
]
```

## Consent Management (GDPR)

In order to enable you must to set **hbAMS.settings.gdpr = true** and add CMP script on your page. A working demo can be found [here](https://ams-hb.firebaseapp.com/demos/cmp.html).

## Custom namespacing
Build prebid from source setting `globalVarName` from package.json to pbams, see [prebid/Prebid.js#409](https://github.com/prebid/Prebid.js/pull/409) for details

# Call Ad Unit through DFP

```
<script type="text/javascript" src="//admytests.firebaseapp.com/js/DAC.js"></script>
<script type="text/javascript" src="//admytests.firebaseapp.com/js/hb_cpprg.js"></script>
<script type="text/javascript" src="//admytests.firebaseapp.com/js/hb_ams-lib.js"></script>
<div id="6544251" class="ams-ad"></div>
```

# Adjust Bid Price for Gross/Net
Bidders may have different pricing deals with publishers, and the returned bid prices may or may not reflect what the publisher will truly receive in the end.
Prebid.js allows the publisher to [adjust the bidding price](http://prebid.org/blog/adjust-bid-price) before the bids targeting are set on the ad server tag. This feature is supported by [pbjs.bidderSettings](http://prebid.org/dev-docs/publisher-api-reference.html#module_pbjs.bidderSettings)

```
    pbjs.bidderSettings = {
      rubicon: {
        bidCpmAdjustment : function(bidCpm){
          // adjust the bid in real time before the auction takes place
          return bidCpm * 0.85;
        }
      }
    };
```

# Tips for Troubleshooting

Check form prebid.org
- [Tips for troubleshooting](http://prebid.org/dev-docs/toubleshooting-tips.html)
- [Prebid.js Troubleshooting Guide](http://prebid.org/dev-docs/prebid-troubleshooting-guide.html)

# Furthermore

- [Implementing OAth's CMP js](https://b2b.oath.com/c/cmp-guide-final?x=XzWzWi)

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.4.2"></a>
## [1.4.2](https://github.com/meumobi/ams-hb/compare/v1.4.1...v1.4.2) (2018-08-15)


### Bug Fixes

* Closes [#29](https://github.com/meumobi/ams-hb/issues/29), add sizeConfig parameter to the pbjs.setConfig ([5a037a1](https://github.com/meumobi/ams-hb/commit/5a037a1))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/meumobi/ams-hb/compare/v1.4.0...v1.4.1) (2018-08-10)


### Bug Fixes

* Closes [#28](https://github.com/meumobi/ams-hb/issues/28), Calls to adserver are not made when no bidders responded ([bc32f37](https://github.com/meumobi/ams-hb/commit/bc32f37))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/meumobi/ams-hb/compare/v1.3.1...v1.4.0) (2018-07-03)


### Bug Fixes

* add basedir on browsersync config access by default on www as public dir ([313d575](https://github.com/meumobi/ams-hb/commit/313d575))
* add ga module on prebid.js ([2b93d49](https://github.com/meumobi/ams-hb/commit/2b93d49))
* allow to track analytics among another trackers ids ([9b08da0](https://github.com/meumobi/ams-hb/commit/9b08da0))
* move global name from pbjs to pbams ([a933137](https://github.com/meumobi/ams-hb/commit/a933137))
* remove ga error tracking ([1b78051](https://github.com/meumobi/ams-hb/commit/1b78051))


### Features

* new ams lib release ([46351dc](https://github.com/meumobi/ams-hb/commit/46351dc))



<a name="1.3.1"></a>
## [1.3.1](https://github.com/meumobi/ams-hb/compare/v1.1.0...v1.3.1) (2018-06-19)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/meumobi/ams-hb/compare/v1.0.0...v1.1.0) (2018-06-19)


### Features

* add cmp demo ([4bd0003](https://github.com/meumobi/ams-hb/commit/4bd0003))
* add gdpr param on settings and add new mango adunits ([4bb8aea](https://github.com/meumobi/ams-hb/commit/4bb8aea))
* add prebid v1.13 and v1.14 ([53837e1](https://github.com/meumobi/ams-hb/commit/53837e1))
* new hb-ams-lib releases ([eb072e8](https://github.com/meumobi/ams-hb/commit/eb072e8))
* update DAC from 3.1.0 to 3.1.1 ([7078352](https://github.com/meumobi/ams-hb/commit/7078352))


### Performance Improvements

* migrate prebid.js to a cdn ([a58e28f](https://github.com/meumobi/ams-hb/commit/a58e28f))



<a name="1.0.0"></a>
# 1.0.0 (2018-06-04)


### Bug Fixes

* Closes [#26](https://github.com/meumobi/ams-hb/issues/26), DOMContentLoaded is requesting bids already requested by prebid ([56a8749](https://github.com/meumobi/ams-hb/commit/56a8749))


### Features

* Prebid.js upgraded to v1.11 BREAKING CHANGES ([7466496](https://github.com/meumobi/ams-hb/commit/7466496))
* Upgrade DAC.js to 3.1.0 - who supports GDPR ([570a8eb](https://github.com/meumobi/ams-hb/commit/570a8eb))


### BREAKING CHANGES

* update all adUnits id from integer to string, 1234 to "1234"


www/js/hb_mas.js
var hbAMS = {
    settings: {
        siteId: '670202',
        prebidAdUnitIds: ["6489219", "6490489"],
        autoRefresh: {
            interval: 15000, // milliseconds
            minVisibility: 0.49, // range 0-1
            onlyIfBidWinner: false
        },
        gdpr: true,
        sizes: {
            desktop: [
                [970, 250],
                [970, 90],
                [728, 90],
                [300, 600],
                [300, 250]
            ],
            tablet: [
                [728, 90],
                [300, 600],
                [160, 600],
                [300, 250]
            ],
            phone: [
                [320, 100],
                [320, 50],
                [300, 250]
            ]
        }
    }
};

var adServer = {};

adServer.config = {
    protocol: 'https',
    server: 'secserv.adtech.de',
    network: '1502.1',
    siteid: '670202',
    gdpr: hbAMS.settings.gdpr,
    params: {
        loc: '100'
    }
};

hbAMS.settings.adUnits = [
    {
        code: "6489219",
        fif: { usefif: true },
        sizes: [[728, 90], [970, 90], [970, 250]],
        bounds: [
            { id: 6489219, min: 769, max: 9999 },
        ],
        sizeid: '225',
        labelAny: ["desktop"],
        bids: [
            {
                bidder: 'smartadserver',
                params: {
                    domain: '//prg.smartadserver.com',
                    siteId: '170999',
                    pageId: '842325',
                    formatId: '45846'
                }
            },
            {
                bidder: 'aol',
                params: {
                    placement: '6033655',
                    network: '1502.1',
                    server: 'adserver.adtech.de',
                }
            },
            {
                bidder: 'rubicon',
                params: {
                    accountId: '14794',
                    siteId: '83734',
                    zoneId: '395240'
                }
            }, 
            {
                bidder: "criteo",
                params: {
                    zoneId: '1119285'
                }
            },
            {
                bidder: "improvedigital",
                params: {
                    placementId: '1167201',
                }
            },
        ]
    },
    {
        code: "6493810",
        fif: { usefif: true },
        sizes: [[320, 50], [320, 100]],
        labelAny: ["phone"],
        bounds: [
            { id: 6493810, min: 0, max: 768 },
        ],
        bids: [
            {
                bidder: 'smartadserver',
                params: {
                    domain: '//prg.smartadserver.com',
                    siteId: '170999',
                    pageId: '845267',
                    formatId: '45846'
                }
            },
            {
                bidder: 'aol',
                params: {
                    placement: '6033655',
                    network: '1502.1',
                    server: 'adserver.adtech.de',
                }
            },
            {
                bidder: 'rubicon',
                params: {
                    accountId: '14794',
                    siteId: '83734',
                    zoneId: '395240'
                }
            }, 
            {
                bidder: "improvedigital",
                params: {
                    placementId: '1167201',
                }
            },
            {
                bidder: 'widespace',
                params: {
                    sid: '620f5133-4b47-4bbf-823c-6b4fe293b6e1',
                    currency: 'USD'
                }
            },
        ]
    },
    {
        code: "6490489",
        fif: { usefif: true },
        autoRefresh: true,
        sizes: [[300, 600], [300, 250]],
        labelAny: ["desktop"],
        bounds: [
            { id: 6494025, min: 0, max: 768 },
            { id: 6490489, min: 769, max: 9999 },
        ],
        sizeid: '170',
        bids: [
            {
                bidder: 'smartadserver',
                params: {
                    domain: '//prg.smartadserver.com',
                    siteId: '170999',
                    pageId: '842325',
                    formatId: '45838'
                }
            },
            {
                bidder: 'rubicon',
                params: {
                    accountId: '14794',
                    siteId: '83734',
                    zoneId: '395240'
                }
            },
            {
                bidder: "criteo",
                params: {
                    zoneId: '1119284'
                }
            },
            {
                bidder: "improvedigital",
                params: {
                    placementId: '1167203',
                }
            },
        ]
    },
    {
        code: "6494025",
        fif: { usefif: true },
        autoRefresh: true,
        sizes: [300, 250],
        labelAny: ["phone"],
        bounds: [
            { id: 6494025, min: 0, max: 768 },
        ],
        bids: [
            {
                bidder: 'smartadserver',
                params: {
                    domain: '//prg.smartadserver.com',
                    siteId: '170999',
                    pageId: '842325',
                    formatId: '45838'
                }
            },
            {
                bidder: 'rubicon',
                params: {
                    accountId: '14794',
                    siteId: '83734',
                    zoneId: '395240'
                }
            },
            {
                bidder: "criteo",
                params: {
                    zoneId: '1119284'
                }
            },
            {
                bidder: "improvedigital",
                params: {
                    placementId: '1167203',
                }
            },
        ]
    },
    {
        code: "6495503",
        fif: { usefif: true },
        autoRefresh: true,
        sizes: [[728, 90], [970, 90], [970, 250]],
        labelAny: ["desktop"],
        bounds: [
            { id: 6495504, min: 0, max: 768 },
            { id: 6495503, min: 769, max: 9999 },
        ],
        sizeid: '225',
        bids: [
            {
                bidder: 'smartadserver',
                params: {
                    domain: '//prg.smartadserver.com',
                    siteId: '170999',
                    pageId: '845211',
                    formatId: '45846'
                }
            },
            {
                bidder: 'rubicon',
                params: {
                    accountId: '14794',
                    siteId: '83734',
                    zoneId: '703820'
                }
            },
        ]
    },
    {
        code: "6495504",
        fif: { usefif: true },
        autoRefresh: true,
        sizes: [[320, 50], [320, 100]],
        labelAny: ["desktop"],
        bounds: [
            { id: 6495504, min: 0, max: 768 },
        ],
        sizeid: '225',
        bids: [
            {
                bidder: 'smartadserver',
                params: {
                    domain: '//prg.smartadserver.com',
                    siteId: '170999',
                    pageId: '845211',
                    formatId: '45846'
                }
            },
            {
                bidder: 'rubicon',
                params: {
                    accountId: '14794',
                    siteId: '83734',
                    zoneId: '703820'
                }
            },
        ]
    }
];
